Implement fuzzy finder as default filter for Select/MultiSelect

Baarsgaard opened the pull request:
My initial suggestion on the implementation in #162 turned sour once I ran into issues with your `filtered_options` implementation.

In the end I decided to replace the `Filter<'a, T>` function in favour of a `Scorer<'a, T>` function.
The only difference between the two is the return type being `Option<i64>` still allowing to implement filters as before, but now also allowing to order/rank the options according to the calculated score.

Moving the prompts to the new scoring logic was minimally invasive and works together with the filtered_options logic already implemented!

I added a `reset_cursor` config option which if true, will reset the cursor position to `0` on filter input changes.

Since it's no longer a "Filter" I have renamed Documentation and types where it makes sense.
This is very much up for debate.

TODO 
- ~~Create a feature flag that disables the fuzzy `DEFAULT_SCORER` and enables a naive one that just does a `contains` check.~~

Baarsgaard commented:
Will get to the failing tests tomorrow :P

Baarsgaard commented:
The rustix dependency breaking msrv tests seems to originate from the tmpfile from the editor feature:
```txt
├── tempfile v3.8.0
│   ├── cfg-if v1.0.0
│   ├── fastrand v2.0.0
│   └── rustix v0.38.13
│       ├── bitflags v2.4.0
│       └── linux-raw-sys v0.4.7
```

I'm not sure it's in scope of this PR to fix that?

<details>
<summary>Full cargo tree output</summary>

```txt

➜ cargo tree --all-features
  Downloaded numtoa v0.1.0
  Downloaded termion v2.0.1
  Downloaded 2 crates (32.4 KB) in 0.25s
inquire v0.6.2 (/home/spu/projects/inquire/inquire)
├── bitflags v2.4.0
├── chrono v0.4.30
│   ├── iana-time-zone v0.1.57
│   └── num-traits v0.2.16
│       [build-dependencies]
│       └── autocfg v1.1.0
├── console v0.15.7
│   ├── lazy_static v1.4.0
│   ├── libc v0.2.147
│   └── unicode-width v0.1.10
├── crossterm v0.25.0
│   ├── bitflags v1.3.2
│   ├── libc v0.2.147
│   ├── mio v0.8.8
│   │   ├── libc v0.2.147
│   │   └── log v0.4.20
│   ├── parking_lot v0.12.1
│   │   ├── lock_api v0.4.10
│   │   │   └── scopeguard v1.2.0
│   │   │   [build-dependencies]
│   │   │   └── autocfg v1.1.0
│   │   └── parking_lot_core v0.9.8
│   │       ├── cfg-if v1.0.0
│   │       ├── libc v0.2.147
│   │       └── smallvec v1.11.0
│   ├── signal-hook v0.3.17
│   │   ├── libc v0.2.147
│   │   └── signal-hook-registry v1.4.1
│   │       └── libc v0.2.147
│   └── signal-hook-mio v0.2.3
│       ├── libc v0.2.147
│       ├── mio v0.8.8 (*)
│       └── signal-hook v0.3.17 (*)
├── dyn-clone v1.0.13
├── fuzzy-matcher v0.3.7
│   └── thread_local v1.1.7
│       ├── cfg-if v1.0.0
│       └── once_cell v1.18.0
├── newline-converter v0.3.0
│   └── unicode-segmentation v1.10.1
├── once_cell v1.18.0
├── tempfile v3.8.0
│   ├── cfg-if v1.0.0
│   ├── fastrand v2.0.0
│   └── rustix v0.38.13
│       ├── bitflags v2.4.0
│       └── linux-raw-sys v0.4.7
├── termion v2.0.1
│   ├── libc v0.2.147
│   └── numtoa v0.1.0
├── unicode-segmentation v1.10.1
└── unicode-width v0.1.10

inquire-derive v0.6.2 (/home/spu/projects/inquire/inquire-derive)
[dev-dependencies]
└── inquire v0.6.2 (/home/spu/projects/inquire/inquire) (*)
```

</details>

mikaelmello commented on inquire/src/prompts/select/action.rs:
```diff
@@ -50,7 +50,7 @@ impl InnerAction for SelectPromptAction {
             Key::End => Self::MoveToEnd,
 
             key => match InputAction::from_key(key, &()) {
-                Some(action) => Self::FilterInput(action),
```
I think the action itself should still be called *Filter*, scoring feels to me like the implementation detail of how we implement the UI feature that is filter input

mikaelmello replied:
(the new term makes sense everywhere else to me)

Baarsgaard replied:
Yeah, I was unsure on the renaming, I even had to concede to using filter in some of the comments.

I will undo the action rename :+1:

mikaelmello reviewed (commented):
This is a really clean solution to the problem, thanks for the contribution!

Do you have an idea of the impact of pulling in the library you're using? Mostly out of curiosity and to make sure we won't be shooting ourselves in the foot.

mikaelmello commented:
> I'm not sure it's in scope of this PR to fix that?

Will take care of it.

Baarsgaard commented:
> > I'm not sure it's in scope of this PR to fix that?
> 
> Will take care of it.
> 
> Update: Fixed it.

Great, I will rebase then!


> Do you have an idea of the impact of pulling in the library you're using? Mostly out of curiosity and to make sure we won't be shooting ourselves in the foot.

In terms of performance:
Performance is great still, I've been using my fork with lists of 1500 options with instant changes to the options.
I would show an Asciinema cast, but the filtering test was on work related Jira issues..

Weight should not be an issue, though I'm not 100% sure on how to quantify this.
The number of sub-dependencies is minimal and there is no non-Rust code at least.

I will try to update my fork tomorrow!

Baarsgaard commented:
I've applied the changes and rebased as discussed earlier, additionally I updated the changelog to reflect the changes.

Before this is considered final, I want to try and implement [Nucleo](https://www.reddit.com/r/rust/comments/165lp5a/announcing_nucleo_a_fast_fuzzy_matcher_library/) instead.
It was apparently announced last week and is supposedly 5-8 times faster than `fuzzy-matcher`

But from a cursory view at how they are using it in the Helix editor, the implementation would potentially need to change a fair bit.
The easiest solution might be to implement a trait similar to Autocompleter which can hold the `Matcher` and the added `Pattern/Atom` types.
The reason being the `Matcher` can be reused for the entire lifetime of the current prompt, but the `Pattern/Atom` that is parsed from the input needs to be re-parsed on any change to the input.
But parsing anew on every Option in the list would be wasteful.

See: https://github.com/helix-editor/helix/blob/master/helix-term/src/ui/menu.rs#L95-L104

Baarsgaard commented:
I'm a bit stuck on using Nucleo instead as of right now. I'm running into the same challenges as I faced with my initial idea in #162.

I would prefer to not rewrite most of the Select/Multiselect Prompt logic just to shoehorn Nucleo in.

mikaelmello commented on inquire/src/prompts/multiselect/mod.rs:
```diff
@@ -172,6 +193,10 @@ where
 
+    /// Default cursor behaviour on filter input change.
+    /// Defaults to true if 'fuzzy' is enabled.
+    pub const DEFAULT_RESET_CURSOR: bool = cfg!(feature = "fuzzy");
```
Honestly, I kind of prefer what you did here, and I'm ok with resetting cursor being the default behaviour (and not being configurable). It feels like a much better default

Baarsgaard replied:
The goal with that config flag was to change the behaviour as little as possible for people already using the library.
But it was necessary for the fuzzy search to feel good.
But it is now `true` per default and can be overwritten instead.

mikaelmello reviewed (dismissed):
I love it, your work is very appreciated :)

Let's just fix the merge conflicts, and merge this. I also think making reset cursor true the built-in behavior to be a good improvement if you feel like it, regardless just lmk when it's ready to merge and I'll push the button

Baarsgaard commented:
Merge conflicts have been resolved and I defaulted the reset behaviour to true!
You're welcome to merge anytime! :D

mikaelmello commented:
Merged!
